### PR TITLE
Implement minimal Application Connector test

### PR DESCRIPTION
**Pull request by VOID404**
**Description**

This PR implements a minimal optimistic test case for the Central Application Gateway. 
Tests are described in AppCR. Tests should be run by the pipeline implemented as part of [test-infra#5563](https://github.com/kyma-project/test-infra/pull/5563)

**Related issue(s)**

#14654

### CODE REVIEW

File: tests/components/application-connector/Makefile [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+GOPATH ?= $(shell go env GOPATH)
+
+GATEWAY_TEST_IMAGE ?= "$(DOCKER_PUSH_REPOSITORY)$(DOCKER_PUSH_DIRECTORY)/gateway-test:$(DOCKER_TAG)"
+MOCK_APP_IMAGE ?= "$(DOCKER_PUSH_REPOSITORY)$(DOCKER_PUSH_DIRECTORY)/mock-app:$(DOCKER_TAG)"
+
+APP_URL = test-app.test.svc.cluster.local
+
+.PHONY: release test-os test image publish clean image-gateway publish-gateway test-gateway certs image-mock-app publish-mock-app
+
+test: test-gateway
```
Comment (franpog859) [inline tests/components/application-connector/Makefile:14]: Maybe we can prepare it for other components. I mean some renaming from the default `test`, `image`, and other commands into `test-application-gateway`, `build-application-gateway`, and so on

File: tests/components/application-connector/README.md [hunk 2]
```diff
@@ -0,0 +1,2 @@
+# Application Gateway tests
```
Comment (franpog859) [inline tests/components/application-connector/README.md:1]: It's not only the Application Gateway test. It's a whole Application Connector test. At least in theory for now

File: tests/components/application-connector/internal/testkit/test-api/handlers.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+	res := EchoResponse{
+		Method:  r.Method,
+		Body:    body,
+		Headers: r.Header,
+		Query:   r.URL.RawQuery,
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	json.NewEncoder(w).Encode(res) // TODO: Handle error
```
Comment (franpog859) [inline tests/components/application-connector/internal/testkit/test-api/handlers.go:36]: Let's handle the error :D

File: tests/components/application-connector/k8s/gateway-test/templates/basic-cases.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
+apiVersion: applicationconnector.kyma-project.io/v1alpha1
+kind: Application
+metadata:
+  name: test-app
```
Comment (franpog859) [inline tests/components/application-connector/k8s/gateway-test/templates/basic-cases.yml:4]: Shouldn't this Application be named `positive-cases-app` or something like this? Also maybe the file could be renamed to `positive-cases-app.yml`. What do you think? Then the failures would be `negative...` or `failing-cases-app`

File: tests/components/application-connector/k8s/gateway-test/templates/basic-cases.yml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  namespace: "{{ .Values.namespace }}"
+spec:
+  description: This is the test service
+  skipVerify: true
+  labels:
+    app: test-app
+  services:
+    - displayName: always ok
+      name: alwaysok
+      providerDisplayName: Rafał
```
Comment (franpog859) [inline tests/components/application-connector/k8s/gateway-test/templates/basic-cases.yml:14]: I'm not sure about it :D

File: tests/components/application-connector/k8s/gateway-test/templates/test-app.yml [hunk 6]
```diff
@@ -0,0 +1,42 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    app: test-app
```
Comment (franpog859) [inline tests/components/application-connector/k8s/gateway-test/templates/test-app.yml:5]: Let's rename it to something like `mock-service` or `mock-application`

File: tests/components/application-connector/k8s/gateway-test/templates/test.yml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: connectivity-test
```
Comment (franpog859) [inline tests/components/application-connector/k8s/gateway-test/templates/test.yml:4]: Maybe `application-connector-test`?

**Review by franpog859 — COMMENTED** [review]
I'd rename `Dockerfile.gateway` into `Dockerfile.application-gateway-test`, and `Dockerfile.testapp` into `Dockerfile.application-gateway-test-mock-service` or `Dockerfile.application-gateway-test-mock-application`

File: tests/components/application-connector/Makefile (hunk 1 above)
Comment (VOID404) [reply tests/components/application-connector/Makefile:14]: We have `image`, depending on `image-gateway`. IMO that's the good way to do it - subtasks for specific components, and a single task for easy running all

File: tests/components/application-connector/k8s/gateway-test/templates/basic-cases.yml (hunk 4 above)
Comment (VOID404) [reply tests/components/application-connector/k8s/gateway-test/templates/basic-cases.yml:4]: Considering the response codes can be stored in the AppCR, we could just have basic cases and merge failing (with wrong secrets) cases into this - the only difference in tests is the response code

File: tests/components/application-connector/k8s/gateway-test/templates/test.yml (hunk 7 above)
Comment (koala7659) [reply tests/components/application-connector/k8s/gateway-test/templates/test.yml:4]: Chaning it to application-gateway-test

File: tests/components/application-connector/k8s/gateway-test/templates/test-app.yml (hunk 6 above)
Comment (koala7659) [reply tests/components/application-connector/k8s/gateway-test/templates/test-app.yml:5]: Will change it to mock-application

File: tests/components/application-connector/k8s/gateway-test/templates/_helpers.tpl [hunk 8]
```diff
@@ -0,0 +1,27 @@
+{{/*
+Expand the name of the chart.
+*/}}
+{{- define "name" -}}
```
Comment (akgalwas) [inline tests/components/application-connector/k8s/gateway-test/templates/_helpers.tpl:4]: name and fullname templates are not used

File: tests/components/application-connector/cmd/test-gateway-app/server.go [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+	basicAuthCredentials := test_api.BasicAuthCredentials{User: cfg.BasicAuthUser, Password: cfg.BasicAuthPassword}
+
+	router := test_api.SetupRoutes(os.Stdout, basicAuthCredentials)
+
+	address := fmt.Sprintf(":%d", cfg.Port)
+	log.Fatal(http.ListenAndServe(address, router))
+}
+
+func newMTLSServer(caCertPath, address string, handler http.Handler) *http.Server {
```
Comment (akgalwas) [inline tests/components/application-connector/cmd/test-gateway-app/server.go:41]: This function is not used.

File: tests/components/application-connector/k8s/gateway-test/templates/_helpers.tpl (hunk 8 above)
Comment (koala7659) [reply tests/components/application-connector/k8s/gateway-test/templates/_helpers.tpl:4]: Removed

**Review by koala7659 — CHANGES_REQUESTED** [review]
When changed in application address of targetUrl to non existent - the tests have crashed:

targetUrl: "http://xxxx.{{ .Values.namespace  }}.svc.cluster.local:8080/v1/api/unsecure/ok"

```
--- FAIL: TestGatewaySuite (0.41s)                                                                                                                                                                            │ delete mode 100644 prow/jobs/kyma/tests/application-gateway-legacy-tests/application-gateway-leg    --- FAIL: TestGatewaySuite/TestSimpleCases (0.41s)                                                                                                                                                        │acy-tests-generic.yaml
        --- PASS: TestGatewaySuite/TestSimpleCases/always_ok (0.36s)                                                                                                                                          │ delete mode 100644 prow/jobs/kyma/tests/application-gateway-tests/application-gateway-tests-gene        --- FAIL: TestGatewaySuite/TestSimpleCases/basic (0.03s)                                                                                                                                              │ric.yaml
panic: runtime error: invalid memory address or nil pointer dereference [recovered]                                                                                                                           │ delete mode 100644 prow/jobs/kyma/tests/rafter/rafter-tests-generic.yaml
        panic: runtime error: invalid memory address or nil pointer dereference                                                                                                                               │ delete mode 100644 prow/jobs/rafter/rafter.yaml
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x11bd182]                                                                                                                                      │ delete mode 100755 prow/scripts/cluster-integration/minio/azure-cleaner.sh
                                                                                                                                                                                                              │ delete mode 100755 prow/scripts/cluster-integration/minio/azure-gateway.sh
goroutine 26 [running]:                                                                                                                                                                                       │ delete mode 100755 prow/scripts/cluster-integration/minio/gcs-gateway.sh
testing.tRunner.func1.2({0x12a6c20, 0x1f1fee0})                                                                                                                                                               │ delete mode 100644 templates/data/rafter-data.yaml
        /usr/local/go/src/testing/testing.go:1389 +0x24e                                                                                                                                                      │     ~/G/s/github.c/kyma-p/test-infra     app-connector-pipelines 
testing.tRunner.func1()                                                                                                                                                                                       │     ~/G/s/github.c/kyma-p/test-infra     app-connector-pipelines                                 /usr/local/go/src/testing/testing.go:1392 +0x39f                                                                                                                                                      │      
panic({0x12a6c20, 0x1f1fee0})                                                                                                                                                                                 │     ~/G/s/github.c/kyma-p/test-infra     app-connector-pipelines  cd $GOPATH/                    /usr/local/go/src/runtime/panic.go:838 +0x207                                                                                                                                                         │     ~/GoPath  ls                                             ✔  k3d-kyma ⎈  15:25:36 
github.com/kyma-project/kyma/tests/components/application-connector/gateway_test.(*GatewaySuite).TestSimpleCases.func1()                                                                                      │bin pkg src
        /connectivity-test/gateway/simple_test.go:26 +0x1e2                                                                                                                                                   │     ~/GoPath  cd bin                                         ✔  k3d-kyma ⎈  15:25:38 
github.com/stretchr/testify/suite.(*Suite).Run.func1(0xc00041e000)                                                                                                                                            │     ~/G/bin  ls                                              ✔  k3d-kyma ⎈  15:25:41 
        /go/pkg/mod/github.com/stretchr/testify@v1.7.1/suite/suite.go:77 +0xf2                                                                                                                                │client-gen      defaulter-gen   go-junit-report informer-gen    present
testing.tRunner(0xc00041e000, 0xc000118048)                                                                                                                                                                   │cmd             depth           goimports       lister-gen      setup-envtest
        /usr/local/go/src/testing/testing.go:1439 +0x102                                                                                                                                                      │deepcopy-gen    dlv             golint          mockery
created by testing.(*T).Run                                                                                                                                                                                   │     ~/G/bin                                                  ✔  k3d-kyma ⎈  15:25:42 
        /usr/local/go/src/testing/testing.go:1486 +0x35f
```

File: tests/components/application-connector/cmd/test-gateway-app/server.go (hunk 9 above)
Comment (koala7659) [reply tests/components/application-connector/cmd/test-gateway-app/server.go:41]: Removed

File: tests/components/application-connector/k8s/gateway-test/templates/basic-cases.yml (hunk 4 above)
Comment (franpog859) [reply tests/components/application-connector/k8s/gateway-test/templates/basic-cases.yml:4]: Ok, it's fine. I was biased by the previous idea :D

File: tests/components/application-connector/README.md (hunk 2 above)
Comment (koala7659) [reply tests/components/application-connector/README.md:1]: I will write here some docs how to run it

File: tests/components/application-connector/k8s/gateway-test/templates/basic-cases.yml (hunk 5 above)
Comment (koala7659) [reply tests/components/application-connector/k8s/gateway-test/templates/basic-cases.yml]: Fixed

File: tests/components/application-connector/Makefile [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+	@echo ""
+	@echo "Gateway test results:"
+ifeq ($(EXPORT_RESULT), true)
+	kubectl -n $(NAMESPACE) logs pod/application-gateway-test -f | tee /dev/stderr | $(GOPATH)/bin/go-junit-report -set-exit-code > junit-report.xml
+else
+	kubectl -n $(NAMESPACE) logs pod/application-gateway-test
+endif
+
+
+clean-gateway:
```
[the remark this answers is not in the document]
Comment (koala7659) [reply tests/components/application-connector/Makefile]: Updated

File: tests/components/application-connector/internal/testkit/test-api/handlers.go (hunk 3 above)
Comment (koala7659) [reply tests/components/application-connector/internal/testkit/test-api/handlers.go]: Error is handled now!

**Review by koala7659 — DISMISSED** [review]
LGTM

**Review by majakurcius — DISMISSED** [review]

File: tests/components/application-connector/Makefile [hunk 11]
```diff
@@ -1,45 +1,47 @@
...
 test-os: test-gateway
 test-skr: test-gateway
 
 test-gateway:
 	@echo $(GOPATH)/bin/go-junit-report --help
 	kubectl create namespace $(NAMESPACE) --dry-run=client -o yaml | kubectl apply -f -
-	helm template k8s/gateway-test --set gatewayTestImage=$(GATEWAY_TEST_IMAGE),mockImage=$(MOCK_APP_IMAGE),namespace=$(NAMESPACE) | kubectl apply -f -
-	sleep 180
-	@echo ""
+	helm template k8s/gateway-test | kubectl apply -f -
```
Comment (VOID404) [inline tests/components/application-connector/Makefile:20]: Values in `values.yaml` have to be in sync with values when pushing, removing variables will make it a bit harder to keep those in sync, but that's only a problem for local development, which is not really a priority.

File: tests/components/application-connector/wait-for-gateway-tests-finish.sh [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+status=$(kubectl get pod/application-gateway-test -n $NAMESPACE -ojsonpath='{ .status.phase}')
+until [[ $status == Succeeded || $status == Failed ]]; do
+  printf '.'
+  sleep 5
+  let "elapsedSeconds=elapsedSeconds+5"
+  if [[ $elapsedSeconds -gt $TIMEOUT_IN_SECONDS ]] ; then
+    exit 0
+  fi
+  status=$(kubectl get pod/application-gateway-test -n $NAMESPACE -ojsonpath='{ .status.phase}')
+  echo $status
```
Comment (VOID404) [inline tests/components/application-connector/wait-for-gateway-tests-finish.sh:18]: maybe exit from the script with the given status?

File: tests/components/application-connector/Makefile [hunk 13]
```diff
@@ -1,45 +1,47 @@
-NAMESPACE ?= "test"
-
-DOCKER_TAG ?= "PR-14743"
-DOCKER_PUSH_REPOSITORY ?= "eu.gcr.io/kyma-project"
+NAMESPACE = test
```
Comment (VOID404) [inline tests/components/application-connector/Makefile:1]:
```suggestion
NAMESPACE ?= test
```
Why do we disallow changing namespace with an env variable?

File: tests/components/application-connector/gateway/simple_test.go [hunk 14]
```diff
@@ -23,11 +23,13 @@ func (gs *GatewaySuite) TestSimpleCases() {
 				gs.T().Log("Calling", entry.CentralGatewayUrl)
 				res, err := http.Get(entry.CentralGatewayUrl)
 				gs.Nil(err)
```
Comment (VOID404) [inline tests/components/application-connector/gateway/simple_test.go:25]: We could make this assert instantly fail the test, which would let us get rid of that nested `if` below

File: tests/components/application-connector/README.md [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+      SKR
+      </summary>
+
+   ```bash
+   kyma deploy --components-file mini-kyma-skr.yaml 
+   ```
+      </details>
+   </div>
+
+4. Run the tests:
```
Comment (NHingerl) [inline tests/components/application-connector/README.md:67]: where's step 3?

**Review by majakurcius — DISMISSED** [review]

File: tests/components/application-connector/wait-for-gateway-tests-finish.sh [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+
+NAMESPACE=test
+TIMEOUT_IN_SECONDS=120
+
+echo "Waiting for tests to finish"
+elapsedSeconds=0
+
+status=$(kubectl get pod/application-gateway-test -n $NAMESPACE -ojsonpath='{ .status.phase}')
+until [[ $status == Succeeded || $status == Failed ]]; do
+  printf '.'
```
Comment (janmedrek) [inline tests/components/application-connector/wait-for-gateway-tests-finish.sh:11]: Do we need to print those statuses? There are no timestamps being printed as well so in my opinion, it does not bring much value.

There is also a `.` printed at the start of each line, do we need it?

File: tests/components/application-connector/README.md [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+   ```bash
+   kyma deploy --components-file mini-kyma-skr.yaml 
+   ```
+      </details>
+   </div>
+
+3. Run the tests:
+
+   ```bash
+   make test-gateway
```
Comment (janmedrek) [inline tests/components/application-connector/README.md:70]: This leaves test leftovers on the cluster, we should either document that cleanup needs to be called or make it a part of the make command that will be used to run the whole flow (preferable).

File: tests/components/application-connector/wait-for-gateway-tests-finish.sh (hunk 16 above)
Comment (koala7659) [reply tests/components/application-connector/wait-for-gateway-tests-finish.sh:11]: The final idea is to leave printing the dot '.' sign every file second to show test progress.  But I have left this intentionally to provide more insight in the logs while running this job on prow only for the initial period after merging the code to main. If we got any problems - this message can show us why the pipeline fail and what is the state of the pod in case of failure.

File: tests/components/application-connector/k8s/gateway-test/templates/test.yml [hunk 18]
```diff
@@ -0,0 +1,17 @@
+apiVersion: v1
+kind: Pod
```
[the remark this answers is not in the document]
Comment (koala7659) [reply tests/components/application-connector/k8s/gateway-test/templates/test.yml:2]: So it was intentional. Using Kubernetes job to run tests was an initial idea for a test runner. However after some time the developers have decided that using a job will bring no value here. 

Regarding `kubectl wait --for=condition=complete` command I have been exploring it for examining pod status for some time, and I could not make it work, so writing a script was the more straightforward way to do the task.

File: tests/components/application-connector/gateway/simple_test.go (hunk 14 above)
Comment (koala7659) [reply tests/components/application-connector/gateway/simple_test.go:25]: Yes, we can do it, the code will be nicer. I suggest to do it in following PR.

File: tests/components/application-connector/Makefile (hunk 13 above)
Comment (koala7659) [reply tests/components/application-connector/Makefile:1]: I agree. Even though it sound as corner case it might be useful for local development. On prow this name will always be `test`.

**Comment by akgalwas** [comment]
It was assumed that we want to:
- store mock app along with tests
- have one go module for test app and tests

I agree that the current structure is not perfect and e can refine it.
My proposal:
image

**Review by franpog859 — APPROVED** [review]
LGTM. Verified locally

**Review by franpog859 — DISMISSED** [review]

**Review by majakurcius — APPROVED** [review]

File: tests/components/application-connector/README.md (hunk 17 above)
Comment (VOID404) [reply tests/components/application-connector/README.md:65]: If we output logs to a file, then there is not much value in not cleaning up the cluster. We could create a new task that runs the current test task and then the cleanup task, changing the default behaviour and letting developers run tests without cleanup.
